File: camerax_android/lib/src/common/permission_manager_impl.dart
import 'package:camerax_android/src/api.dart';
import 'package:camerax_platform_interface/camerax_platform_interface.dart';

final class PermissionManagerImpl implements PermissionManager {
  final PermissionManagerProxyApi api;

  PermissionManagerImpl.internal(this.api);

  @override
  Future<bool> checkPermission(PermissionManager$Permission permission) =>
      api.checkPermissioin(permission.api);

  @override
  Future<bool> requestPermissions(
    List<PermissionManager$Permission> permissions,
  ) => api.requestPermissions(permissions.map((e) => e.api).toList());
}

final class PermissionManagerChannelImpl extends PermissionManagerChannel {
  @override
  PermissionManager create() {
    final api = PermissionManagerProxyApi.instance;
    return PermissionManagerImpl.internal(api);
  }
}

extension PermissionManager$PermissionX on PermissionManager$Permission {
  PermissionManagerPermissionApi get api =>
      PermissionManagerPermissionApi.values[index];
}

File: camerax_android/android/src/main/kotlin/dev/zeekr/camerax_android/common/PermissionManagerImpl.kt
package dev.zeekr.camerax_android.common

import dev.zeekr.camerax_android.CameraXApiPigeonProxyApiRegistrar
import dev.zeekr.camerax_android.PermissionManagerPermissionApi
import dev.zeekr.camerax_android.PigeonApiPermissionManagerProxyApi
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class PermissionManagerImpl(registrar: CameraXApiPigeonProxyApiRegistrar) :
    PigeonApiPermissionManagerProxyApi(registrar) {
    override fun instance(): PermissionManager {
        return PermissionManager
    }

    override fun checkPermissioin(
        pigeon_instance: PermissionManager,
        permission: PermissionManagerPermissionApi
    ): Boolean {
        return PermissionManager.checkPermission(permission.impl)
    }

    override fun requestPermissions(
        pigeon_instance: PermissionManager,
        permissions: List<PermissionManagerPermissionApi>,
        callback: (Result<Boolean>) -> Unit
    ) {
        CoroutineScope(Dispatchers.Main).launch {
            try {
                val permissionImpls = permissions.map { permission -> permission.impl }
                val isGranted = PermissionManager.requestPermissions(permissionImpls)
                callback(Result.success(isGranted))
            } catch (e: Exception) {
                callback(Result.failure(e))
            }
        }
    }
}

val PermissionManagerPermissionApi.impl: PermissionManager.Permission
    get() = when (this) {
        PermissionManagerPermissionApi.VIDEO -> PermissionManager.Permission.VIDEO
        PermissionManagerPermissionApi.AUDIO -> PermissionManager.Permission.AUDIO
    }